src/realizations/catchment/Bmi_C_Adapter.cpp

```diff
@@ -0,0 +1,423 @@
+    int result = bmi_model->get_start_time(bmi_model.get(), &start_time);
+    if (result != BMI_SUCCESS) {
+        throw std::runtime_error(model_name + " failed to get model start time.");
+    }
+    return start_time;
+}
+
+/**
+ * Get the backing model "next" time step size.
+ *
+ * Get the time step size used in the backing model for the "next", in the units expressed by
+ * `GetTimeUnits()`.
+ *
+ * This type does not make any assumptions about how the backing model handles time steps, including whether
+ * the time step size is fixed.  This creates some complexities.  In particular, this function makes a
+ * best-effort for determining the time step size.  It attempts to do this first by analyzing the data in
+ * the forcing file at `forcing_file_path`, for whatever the next time step is.  This may not be possible,
+ * in which case it falls back to the value returned from the analogous BMI method from the backing model.
+ *
+ * @return The time step size of the model.
+ */
+double Bmi_C_Adapter::GetTimeStep() {
+    // FIXME: finish implementing properly beyond just reading from model, as noted in docstring above
+    double current_time = GetCurrentTime();
+    double start_time = GetStartTime();
+    std::vector<int> ts_sizes = GetValue<int>("TIME_STEP_DELTA");
+    int i = 0;
+    while (i < ts_sizes.size() && current_time > start_time)
+        current_time -= ts_sizes[i++];
+    return (i >= ts_sizes.size()) ? ts_sizes[i-1] : ts_sizes[i];
+}
+
+std::string Bmi_C_Adapter::GetTimeUnits() {
+    char time_units_cstr[BMI_MAX_UNITS_NAME];
+    int result = bmi_model->get_time_units(bmi_model.get(), time_units_cstr);
+    if (result != BMI_SUCCESS) {
+        throw std::runtime_error(model_name + " failed to read time units from model.");
+    }
+    return std::string(time_units_cstr);
+}
+
+int Bmi_C_Adapter::GetVarItemsize(std::string name) {
+    int size;
+    int success = bmi_model->get_var_itemsize(bmi_model.get(), name.c_str(), &size);
+    if (success != BMI_SUCCESS) {
+        throw std::runtime_error(model_name + " failed to get variable item size for " + name + ".");
+    }
+    return size;
+}
+
+int Bmi_C_Adapter::GetVarNbytes(std::string name) {
+    int size;
+    int success = bmi_model->get_var_nbytes(bmi_model.get(), name.c_str(), &size);
+    if (success != BMI_SUCCESS) {
+        throw std::runtime_error(model_name + " failed to get variable array size (i.e., nbytes) for " + name + ".");
+    }
+    return size;
+}
+
+std::string Bmi_C_Adapter::GetVarType(std::string name) {
+    char type_c_str[BMI_MAX_TYPE_NAME];
+    int success = bmi_model->get_var_type(bmi_model.get(), name.c_str(), type_c_str);
+    if (success != BMI_SUCCESS) {
+        throw std::runtime_error(model_name + " failed to get variable type for " + name + ".");
+    }
+    std::string type_str(type_c_str);
+    return type_str;
+}
+
+std::string Bmi_C_Adapter::GetVarUnits(std::string name) {
+    char units_c_str[BMI_MAX_TYPE_NAME];
```

Comment:
This comes from the externally obtained _bmi.h_ file and is defined by the spec.  Assuming the internal copy we are using remains consistent with the supported BMI version of any particular model, your scenario shouldn't happen.  

Of course, it could if the model wasn't implemented correctly.  But the way NGen should handle incorrectly implemented models in various situations, including whether NGen should account for that at all, is a discussion I think needs to happen separately.